### ISSUE TITLE
ASAN performance on exit()

### ISSUE THREAD

**Issue by geremycondra** [google/sanitizers#1227]
Hello, another performance observation related to https://github.com/google/sanitizers/issues/1226.

The TL;DR is that the same synthetic workload appears to be spending about 10x as much time in exit() and friends under ASAN as not. I'm wondering if there's a way to avoid that. I've attached some flamegraphs as well as a synthetic workload to help demonstrate what I'm on about.

Those images were generated by running ASAN and non-ASAN versions of the same program under:

```
perf record -F99 --call-graph dwarf -- <progname>
perf script | stackcollapse-perf.pl > out.perf-folded;
cat out.perf-folded | flamegraph.pl > "<progname>_perf.svg";
```
where the perl scripts referenced above are those from https://github.com/brendangregg/FlameGraph.

Here is the program:
```
#include <sys/types.h>
#include <unistd.h>
#include <stdlib.h>

int dothething(void) {
	for (int i=0; i < 100000; i++) {
		pid_t pid = fork();
		if (!pid) {
			exit(0);
		}
	}
	return 0;
}

int main() {
	dothething();
	exit(0);
}
```
and the relevant makefile

```
FLAGS_PROF = -fprofile-arcs -ftest-coverage
FLAGS_ASAN = -fsanitize=address

all:
	gcc $(FLAGS_PROF) $(FLAGS_ASAN) forktest.c -o forktest_prof_asan && mv forktest.gcno forktest_prof_asan.gcno
	gcc $(FLAGS_PROF) forktest.c -o forktest_prof && mv forktest.gcno forktest_prof.gcno
	gcc $(FLAGS_ASAN) forktest.c -o forktest_asan
	gcc forktest.c -o forktest

clean:
	rm -rf forktest forktest_* *.gcda *.gcno default.profraw
```

The prof versions are also generated along with flame graphs for comparison with https://github.com/google/sanitizers/issues/1226, but would seem to be a separate question.

[fwdinterestingperfresultsforasangcov.zip](https://github.com/google/sanitizers/files/4483789/fwdinterestingperfresultsforasangcov.zip)

**Comment by yugr** [comment]
Could you try running with `export ASAN_OPTIONS=detect_leaks=0`?

**Comment by geremycondra** [comment]
Yep, that works. Thanks.
